File: PSPageControl.m
//
//  PSPageControl.m
//  PSPageControl
//
//  Created by Piotr Sochalewski on 14.08.2015.
//  Copyright (c) 2015 Piotr Sochalewski. All rights reserved.
//

#import "PSPageControl.h"

@interface PSPageControl ()

#pragma mark - Properties
#pragma mark Main properties
@property (strong, nonatomic) NSArray *subviewsNotAllowedToRemoveFromSuperview;
@property (strong, nonatomic) UIImageView *background;
@property (strong, nonatomic) UIPageControl *pageControl;

#pragma mark Touch properties
@property (assign, nonatomic) CGPoint touchPosition;
@property (assign, nonatomic) NSInteger currentViewIndex;
@property (assign, nonatomic) CGPoint backgroundLayerFrameOrigin;

@end

@implementation PSPageControl

#pragma mark - Init

- (instancetype)init {
    self = [super init];
    if (self) {
        [self initialize];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initialize];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initialize];
    }
    return self;
}

- (void)initialize {
    // Default values
    self.offsetPerPageInPixels = 40;
    self.currentViewIndex = 0;
    
    // Background image
    self.background = [UIImageView new];
    self.background.contentMode = UIViewContentModeLeft;
    [self addSubview:self.background];
    
    // Page control
    self.pageControl = [UIPageControl new];
    self.pageControl.numberOfPages = self.views.count;
    [self.pageControl addTarget:self action:@selector(pageControlValueChanged:) forControlEvents:UIControlEventValueChanged];
    [self addSubview:self.pageControl];
    
    // Array of views not allowed to remove from superview
    self.subviewsNotAllowedToRemoveFromSuperview = @[self.background, self.pageControl];
}

#pragma mark - Layout

- (void)layoutSubviews {
    [super layoutSubviews];
    
    // Backround image
    self.background.frame = self.frame;
    self.background.layer.frame = CGRectMake(-(NSInteger)self.offsetPerPageInPixels, 0, CGRectGetWidth(self.background.layer.frame), CGRectGetHeight(self.background.layer.frame));
    self.backgroundLayerFrameOrigin = self.background.layer.frame.origin;
    
    //Page control
    self.pageControl.frame = CGRectMake(10, CGRectGetHeight(self.frame)-50, CGRectGetWidth(self.frame)-20, 40);
}

#pragma mark - Setters

- (void)setBackgroundPicture:(UIImage *)backgroundPicture {
    _backgroundPicture = backgroundPicture;
    
    CGSize size = AVMakeRectWithAspectRatioInsideRect(backgroundPicture.size, CGRectMake(0, 0, CGFLOAT_MAX, CGRectGetHeight(self.frame))).size;
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0);
    [backgroundPicture drawInRect:CGRectMake(0, 0, size.width, size.height)];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    self.background.image = image;
}

- (void)setViews:(NSArray *)views {
    _views = views;
    
    for (UIView *subview in self.subviews) {
        if (![self.subviewsNotAllowedToRemoveFromSuperview containsObject:subview]) {
            [subview removeFromSuperview];
        }
    }
    
    for (NSUInteger i = 0; i < views.count; i++) {
        UIView *view = (UIView *)views[i];
        view.frame = CGRectMake(i*CGRectGetWidth(self.frame), 0, CGRectGetWidth(self.frame), CGRectGetHeight(self.frame));
        [self addSubview:view];
    }
    
    [self bringSubviewToFront:self.pageControl];
    self.pageControl.numberOfPages = self.views.count;
}

- (void)setPageIndicatorTintColor:(UIColor *)pageIndicatorTintColor {
    _pageIndicatorTintColor = pageIndicatorTintColor;
    self.pageControl.pageIndicatorTintColor = pageIndicatorTintColor;
}

- (void)setCurrentPageIndicatorTintColor:(UIColor *)currentPageIndicatorTintColor {
    _currentPageIndicatorTintColor = currentPageIndicatorTintColor;
    self.pageControl.currentPageIndicatorTintColor = currentPageIndicatorTintColor;
}

#pragma mark - Views

- (void)setVisibleViewWithIndex:(NSInteger)index setCurrentPage:(BOOL)setCurrentPage {
    // Background image
    if (index != self.currentViewIndex) {
        CGFloat newX = (index > self.currentViewIndex) ? self.backgroundLayerFrameOrigin.x-self.offsetPerPageInPixels : self.backgroundLayerFrameOrigin.x+self.offsetPerPageInPixels;
        self.backgroundLayerFrameOrigin = CGPointMake(newX, 0);
    }
    
    [UIView animateWithDuration:(index == self.currentViewIndex) ? 0.3 : 0.2 animations:^{
        self.background.layer.frame = CGRectMake(self.backgroundLayerFrameOrigin.x, 0, CGRectGetWidth(self.background.layer.frame), CGRectGetHeight(self.background.layer.frame));
    }];
    
    // Views
    [UIView animateWithDuration:0.2 animations:^{
        // Center (show) view with current index
        UIView *view = (UIView *)self.views[index];
        view.frame = CGRectMake(0, 0, CGRectGetWidth(self.frame), CGRectGetHeight(self.frame));
        
        // Move to the left views with index lower than >index<
        for (NSInteger i = 0; i < index; i++) {
            UIView *view = (UIView *)self.views[i];
            view.frame = CGRectMake(-(index-i)*CGRectGetWidth(self.frame), 0, CGRectGetWidth(view.frame), CGRectGetHeight(view.frame));
        }
        
        // Move to the right view with index higher than >index<
        for (NSInteger i = index+1; i < self.views.count; i++) {
            UIView *view = (UIView *)self.views[i];
            view.frame = CGRectMake((i-index)*CGRectGetWidth(self.frame), 0, CGRectGetWidth(view.frame), CGRectGetHeight(view.frame));
        }
    } completion:^(BOOL finished) {
        if (setCurrentPage) {
            self.pageControl.currentPage = index;
        }
        self.currentViewIndex = index;
    }];
}

#pragma mark - Math

- (NSInteger)differenceBetweenStartPosition:(CGPoint)startPosition andCurrentPosition:(CGPoint)currentPosition {
    return (currentPosition.x - startPosition.x);
}

- (NSInteger)differenceFromTouches:(NSSet *)touches {
    CGPoint movingPosition = [touches.anyObject locationInView:self];
    
    return [self differenceBetweenStartPosition:self.touchPosition andCurrentPosition:movingPosition];
}

#pragma mark - Swipes

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesBegan:touches withEvent:event];
    
    self.touchPosition = [touches.anyObject locationInView:self];
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesMoved:touches withEvent:event];
    
    NSInteger differenceInTouchXAxis = [self differenceFromTouches:touches];
    
    [UIView animateWithDuration:0.1 animations:^{
        for (NSInteger i = 0; i < self.views.count; i++) {
            UIView *view = (UIView *)self.views[i];
            view.frame = CGRectMake(((i-self.currentViewIndex)*CGRectGetWidth(view.frame))+differenceInTouchXAxis, CGRectGetMinY(view.frame), CGRectGetWidth(view.frame), CGRectGetHeight(view.frame));
        }
        
        self.background.layer.frame = CGRectMake(self.backgroundLayerFrameOrigin.x + (differenceInTouchXAxis/CGRectGetWidth(self.frame))*self.offsetPerPageInPixels, 0, CGRectGetWidth(self.background.layer.frame), CGRectGetHeight(self.background.layer.frame));
    }];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesEnded:touches withEvent:event];
    
    NSInteger differenceInTouchXAxis = [self differenceFromTouches:touches];
    if (differenceInTouchXAxis < -100) {
        [self setVisibleViewWithIndex:((self.currentViewIndex+1 >= self.views.count) ? self.currentViewIndex : self.currentViewIndex+1) setCurrentPage:YES];
    } else if (differenceInTouchXAxis > 100) {
        [self setVisibleViewWithIndex:((self.currentViewIndex < 1) ? self.currentViewIndex : self.currentViewIndex-1) setCurrentPage:YES];
    } else {
        [self setVisibleViewWithIndex:self.currentViewIndex setCurrentPage:NO];
    }
}

#pragma mark - Page Control

- (void)pageControlValueChanged:(UIPageControl *)pageControl {
    [self setVisibleViewWithIndex:pageControl.currentPage setCurrentPage:NO];
}

@end

File: PSPageControl.swift
//
//  PSPageControl.swift
//  PSPageControl
//
//  Originaly created by Piotr Sochalewski on 14.08.2015.
//  Rewritten to Swift by Piotr Sochalewski on 05.02.2016.
//

import AVFoundation
import UIKit
import UIImageViewAlignedSwift

/**
 Event to detect when PSPageControl was changed
 */
public protocol PSPageControlProtocol {
    func didChange(index: Int)
}

open class PSPageControl: UIView {
    
    /**
     The image shown in the background. It should be horizontal with proper ratio and high resolution.
     */
    open var backgroundPicture: UIImage? {
        didSet {
            guard let backgroundPicture = backgroundPicture else { return }
            
            let size = AVMakeRect(aspectRatio: backgroundPicture.size,
                                  insideRect: CGRect(x: 0.0, y: 0.0, width: CGFloat.greatestFiniteMagnitude, height: frame.height)).size
            UIGraphicsBeginImageContextWithOptions(size, false, 0.0)
            backgroundPicture.draw(in: CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height))
            let image = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()
            
            background.image = image
        }
    }
    
    /**
     The array of `UIView`s to be shown by page control.
     */
    open var views: [UIView]? {
        didSet {
            subviews.filter( { !subviewsNotAllowedToRemoveFromSuperview.contains($0) } ).forEach {
                $0.removeFromSuperview()
            }
            
            guard let views = views else { pageControl.numberOfPages = 0; return }
            
            for (index, view) in views.enumerated() {
                view.frame = CGRect(x: CGFloat(index) * frame.width,
                    y: 0.0,
                    width: frame.width,
                    height: frame.height)
                addSubview(view)
            }
            
            pageControl.numberOfPages = views.count
        }
    }
    
    /**
     Offset per page in pixels. Default is `40`.
     */
    open var offsetPerPage: UInt = 40
    
    /**
     The tint color to be used for the page indicator.
     */
    open var pageIndicatorTintColor: UIColor? {
        set {
            pageControl.pageIndicatorTintColor = newValue
        }
        get {
            return self.pageIndicatorTintColor
        }
    }
    
    /**
     The tint color to be used for the current page indicator.
     */
    open var currentPageIndicatorTintColor: UIColor? {
        set {
            pageControl.currentPageIndicatorTintColor = newValue
        }
        get {
            return self.currentPageIndicatorTintColor
        }
    }
    
    /**
     The frame rectangle, which describes the page indicators' location and size in its superview’s coordinate system.
     
     Changes to this property can be animated.
     */
    open var pageIndicatorFrame: CGRect {
        get {
            return pageControl.frame
        }
        set {
            pageControl.frame = newValue
        }
    }
    
    fileprivate var subviewsNotAllowedToRemoveFromSuperview = [UIView]()
    fileprivate var background = UIImageViewAligned()
    fileprivate var pageControl = UIPageControl()
    fileprivate var touchPosition: CGPoint?
    fileprivate var backgroundLayerFrameOrigin: CGPoint?
    
    /// Delegate to detect when current PSPageControl changed
    open var delegate: PSPageControlProtocol?

    /**
         Get current PSPageViewControl
     */
    open var currentViewIndex = 0 {
        didSet {
            delegate?.didChange(index: currentViewIndex)
        }
    }
    
    
    fileprivate func setup() {
        // Background image
        background.contentMode = .scaleAspectFill
        background.alignment = .left
        addSubview(background)
        
        // Page control
        addSubview(pageControl)
        
        // Array of views not allowed to remove from superview
        subviewsNotAllowedToRemoveFromSuperview = [background, pageControl]
    }
    
    public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        
        // Background image
        background.frame = frame
        background.layer.frame = CGRect(x: CGFloat(-currentViewIndex - 1) * CGFloat(offsetPerPage), y: 0.0, width: background.layer.frame.width, height: background.layer.frame.height)
        backgroundLayerFrameOrigin = background.layer.frame.origin
        
        // Page control
        pageControl.frame = CGRect(x: 10.0, y: frame.height - 50.0, width: frame.width - 20.0, height: 40.0)
    }
    
    // MARK: - Views
    
    /**
     Shows a view with a given index.
     
     - parameter index: The index of page that should be shown by the receiver as a white dot. The property value is an integer specifying a page shown minus one; thus a value of indicates the first page. Values outside the possible range are pinned to either 0 or numberOfPages minus 1.
     - returns: A Boolean value that determines whether the function finished with success.
     */
    public func showView(withIndex index: Int) -> Bool {
        if let views = views, index < views.count, index >= 0 {
            showView(withIndex: index, setCurrentPage: true)
            return true
        } else {
            return false
        }
    }
    
    fileprivate func showView(withIndex index: Int, setCurrentPage currentPage: Bool) {
        // Background image
        if index != currentViewIndex {
            let newX = CGFloat(-index - 1) * CGFloat(offsetPerPage)
            backgroundLayerFrameOrigin = CGPoint(x: newX, y: 0.0)
        }
        
        let duration = (index == currentViewIndex) ? 0.3 : 0.2
        currentViewIndex = index
        UIView.animate(withDuration: duration) {
            self.background.layer.frame = CGRect(x: self.backgroundLayerFrameOrigin!.x,
                                                 y: 0.0,
                                                 width: self.background.layer.frame.width,
                                                 height: self.background.layer.frame.height)
        }
        
        // Views
        UIView.animate(withDuration: 0.2, animations: {
            // Center (show) view with current index
            let view = self.views![index]
            view.frame = CGRect(x: 0.0, y: 0.0, width: self.frame.width, height: self.frame.height)
            
            // Move to the left views with index lower than >index<
            // and to the right views with index higher than >index<
            for (i, view) in self.views!.enumerated() {
                let newX: CGFloat
                switch i {
                case let x where x < index:
                    newX = CGFloat(-(index - i)) * self.frame.width
                case let x where x > index:
                    newX = CGFloat(i - index) * self.frame.width
                default:
                    newX = 0.0
                }
                
                view.frame.origin = CGPoint(x: newX, y: 0.0)
            }
        }) {  _ in
            if currentPage {
                self.pageControl.currentPage = index
            }
        }
    }
    
    // MARK: - Touches
    
    fileprivate func difference(fromTouches touches: Set<UITouch>) -> Int {
        let movingPosition = touches.first?.location(in: self)
        
        return Int(movingPosition!.x - touchPosition!.x)
    }
    
    open override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        
        touchPosition = touches.first?.location(in: self)
    }
    
    open override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesMoved(touches, with: event)
        
        let differenceInTouchXAxis = difference(fromTouches: touches)
        
        UIView.animate(withDuration: 0.1) {
            for (index, view) in self.views!.enumerated() {
                view.frame = CGRect(x: CGFloat(index - self.currentViewIndex) * view.frame.width + CGFloat(differenceInTouchXAxis), y: 0.0,
                    width: self.background.layer.frame.width, height: self.background.layer.frame.height)
            }
            let x = CGFloat(-self.currentViewIndex - 1) * CGFloat(self.offsetPerPage)
            let deltaX = (CGFloat(differenceInTouchXAxis) / self.frame.width) * CGFloat(self.offsetPerPage)
            self.background.layer.frame = CGRect(x: x + deltaX, y: 0.0, width: self.background.layer.frame.width, height: self.background.layer.frame.height)
        }
    }
    
    open override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        
        let differenceInTouchXAxis = difference(fromTouches: touches)
        
        switch differenceInTouchXAxis {
        case let x where x < -100:
            showView(withIndex: currentViewIndex + 1 >= views!.count ? currentViewIndex : currentViewIndex + 1, setCurrentPage: true)
        case let x where x > 100:
            showView(withIndex: currentViewIndex < 1 ? currentViewIndex : currentViewIndex - 1, setCurrentPage: true)
        default:
            showView(withIndex: currentViewIndex, setCurrentPage: false)
        }
    }
}
